feat: list_code_usages tool

lazymaniac opened the pull request:
## Description

### Why
To provide precise and exact context required by the LLM to solve coding tasks without guesses and assumptions about code fragments. It's specifically designed to solve smaller and direct tasks, and it achieves good results.

### What
This is a new tool for CodeCompanion called `list_code_usages`. Previously I had it in my personal config, but I decided to create a PR as the tool is more or less polished. It works similarly to the `vectorcode` tool, but instead of querying vector databases, it utilizes semantic knowledge of LSP.

### How it works
1. Finds symbols with LSP workspace/symbol function
2. Finds symbols with grep (this supplements LSP results)
3. Goes to the file where the symbol is found and places the cursor at the symbol position
4. Calls LSP methods: 'textDocument/definition', 'textDocument/references', 'textDocument/implementation', 'textDocument/declaration', 'textDocument/typeDefinition', 'textDocument/hover'
5. For all the results from LSP and grep, uses Tree-sitter locals to find the closest most significant code block (function, class, struct, etc.)
6. If Tree-sitter fails, uses fallback to manually find the code block for results

I know it's a little bit bloated and complex compared to other tools and the general philosophy of this plugin, but it's helpful for me and maybe others can benefit from it as well.

## Screenshots

<!-- Add screenshots of the changes if applicable, to help visualize the change. -->
<img width="1728" height="1117" alt="Screenshot 2025-07-23 at 08 01 32" src="https://github.com/user-attachments/assets/e4fd2c3d-231a-4b91-b486-29ad93f9b46a" />
<img width="1728" height="1117" alt="Screenshot 2025-07-23 at 08 01 50" src="https://github.com/user-attachments/assets/35bb7da2-9a6e-4275-8bd6-6959e37d6ac7" />
<img width="1728" height="1117" alt="Screenshot 2025-07-23 at 08 04 38" src="https://github.com/user-attachments/assets/1ece7ac4-5b1a-40fc-9737-f49b186138ac" />

## Checklist

- [x] I've read the [contributing](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md) guidelines and have adhered to them in this PR
- [x] I've updated `CodeCompanion.has` in the [init.lua](https://github.com/olimorris/codecompanion.nvim/blob/main/lua/codecompanion/init.lua#L239) file for my new feature
- [x] I've added [test](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md#testing) coverage for this fix/feature
- [x] I've updated the README and/or relevant docs pages
- [x] I've run `make all` to ensure docs are generated, tests pass and my formatting is applied

olimorris commented:
Thanks @lazymaniac. Very interesting idea and super timely! On the todolist for the next month was for me to add a `list_code_usages` tool, akin to what GitHub Copilot Chat has. FYI, their tool schema is:

```json
"function": {
    "name": "list_code_usages",
    "description": "Request to list all usages (references, definitions, implementations etc) of a function, class, method, variable etc. Use this tool when \n1. Looking for a sample implementation of an interface or class\n2. Checking how a function is used throughout the codebase.\n3. Including and updating all usages when changing a function, method, or constructor",
    "parameters": {
        "type": "object",
        "properties": {
            "symbolName": {
                "type": "string",
                "description": "The name of the symbol, such as a function name, class name, method name, variable name, etc."
            },
            "filePaths": {
                "type": "array",
                "description": "One or more file paths which likely contain the definition of the symbol. For instance the file which declares a class or function. This is optional but will speed up the invocation of this tool and improve the quality of its output.",
                "items": {
                    "type": "string"
                }
            }
        },
        "required": [
            "symbolName"
        ]
    }
}
```

Would be more than happy to merge your PR if you can conform to this schema.

olimorris commented:
@lazymaniac yep exactly like Copilot. I want the tools to be super simple to allow users to have as much success as possible with lesser LLMs. My intention is to ultimately bundle this as part of an agent grouping of tools with an overlaid system prompt.

Having to explain `get_definition`, `get_references`, `get_implementation, `hover` to an LLM is really complex.

lazymaniac commented:
Looks very similar to what I have except of one thing. I extract  encapsulating code block instead of sharing where the code block is. This way the tool is self sufficient and doesn't require other tools to actually get the symbol usages.

lazymaniac commented:
Example tool execution:

## Me

> Context:
> - <tool>list_code_usages</tool>

simplify process_lsp_result

## Anthropic (model=claude-3-7-sonnet-20250219)

### Reasoning

The user is asking me to simplify a function called `process_lsp_result`. However, they haven't provided the code for this function, or any context about what it does or where it's defined.

To help the user effectively, I need to find the definition of `process_lsp_result` in their codebase. I'll use the `list_code_usages` tool to locate this function and understand its implementation.

I'll search for the function name "process_lsp_result" to find its definition and possibly its usages.

### Response

I'll help you simplify the `process_lsp_result` function, but first I need to find its definition in your codebase.

### Tool Output


documentation of symbol: `process_lsp_result`
---
```lua
(method) LSPCaller:process_lsp_result(result: table, operation: any)
  -> result: table
```

---

 Processes LSP results, handling both single items and arrays

@*param* `result` — LSP result data, either single item or array

@*return* `result` — Contains status and data indicating processing result

implementation of symbol: `process_lsp_result`
---
Filename: /Users/sebastian/workspace/codecompanion.nvim/lua/codecompanion/strategies/chat/agents/tools/list_code_usages.lua:340-372
```lua
function LSPCaller:process_lsp_result(result, operation)
  if result.contents then
    if not self.symbol_data[operation] then
      self.symbol_data[operation] = {}
    end
    table.insert(self.symbol_data[operation], {
      code_block = result.contents,
    })
    return { status = "success", data = "Hover content processed" }
  end

  if result.range then
    return self:process_single_range(result.uri or result.targetUri, result.range, operation)
  end

  if #result > 50 then
    return { status = "error", data = "Too many results for symbol operation." }
  end

  local errors = {}
  for _, item in pairs(result) do
    local process_result =
      self:process_single_range(item.uri or item.targetUri, item.range or item.targetSelectionRange, operation)
    if process_result.status == "error" then
      table.insert(errors, process_result.data)
    end
  end

  if #errors > 0 then
    return { status = "error", data = table.concat(errors, "; ") }
  end

  return { status = "success", data = "Results processed" }
end
```

references of symbol: `process_lsp_result`
---
Filename: /Users/sebastian/workspace/codecompanion.nvim/lua/codecompanion/strategies/chat/agents/tools/list_code_usages.lua:384-383
```lua
local process_result = self:process_lsp_result(lsp_results or {}, operation)
```


### Reasoning

Now I understand the function that needs simplification. The `process_lsp_result` function is in the LSPCaller class and handles LSP results from various operations. Let me analyze the current implementation:

```lua
function LSPCaller:process_lsp_result(result, operation)
  if result.contents then
    if not self.symbol_data[operation] then
      self.symbol_data[operation] = {}
    end
    table.insert(self.symbol_data[operation], {
      code_block = result.contents,
    })
    return { status = "success", data = "Hover content processed" }
  end

  if result.range then
    return self:process_single_range(result.uri or result.targetUri, result.range, operation)
  end

  if #result > 50 then
    return { status = "error", data = "Too many results for symbol operation." }
  end

  local errors = {}
  for _, item in pairs(result) do
    local process_result =
      self:process_single_range(item.uri or item.targetUri, item.range or item.targetSelectionRange, operation)
    if process_result.status == "error" then
      table.insert(errors, process_result.data)
    end
  end

  if #errors > 0 then
    return { status = "error", data = table.concat(errors, "; ") }
  end

  return { status = "success", data = "Results processed" }
end
```

The function handles different types of LSP results:
1. Single item with `contents` field (hover content)
2. Single item with `range` field
3. Array of items with ranges

It seems to have some repetitive code and could benefit from simplification. Let's rewrite it:

Simplification approach:
1. Extract the contents handling into a separate function
2. Improve error handling
3. Make the code more concise
4. Better structure the logic

### Response

Based on the code you've shared, I'll simplify the `process_lsp_result` function. Here's a more concise version:

```lua
function LSPCaller:process_lsp_result(result, operation)
  -- Handle hover content
  if result.contents then
    self.symbol_data[operation] = self.symbol_data[operation] or {}
    table.insert(self.symbol_data[operation], { code_block = result.contents })
    return { status = "success", data = "Hover content processed" }
  end

  -- Handle single range result
  if result.range then
    return self:process_single_range(result.uri or result.targetUri, result.range, operation)
  end

  -- Handle array of results
  if #result > 50 then
    return { status = "error", data = "Too many results for symbol operation." }
  end

  -- Process each result item
  local errors = {}
  for _, item in ipairs(result) do
    local uri = item.uri or item.targetUri
    local range = item.range or item.targetSelectionRange
    local process_result = self:process_single_range(uri, range, operation)
    
    if process_result.status == "error" then
      table.insert(errors, process_result.data)
    end
  end

  return #errors > 0
    and { status = "error", data = table.concat(errors, "; ") }
    or { status = "success", data = "Results processed" }
end
```

The main improvements:
- Used `ipairs` instead of `pairs` for array iteration
- Simplified initialization of the symbol_data table
- Improved variable names and extracted them for clarity
- Used a ternary-style return for the final condition
- Maintained the same functionality while reducing code complexity

Would you like me to explain any specific part of this simplification in more detail?

## Me

> Context:
> - <tool>list_code_usages</tool>

olimorris commented:
This is looking so good! very excited to be able to try this out.

One thought that I forgot to mention (and I haven't documented): across CodeCompanion now, all paths that get shared with the LLM are relative to the current working directory. This was because I didn't want to give it knowledge of a users disk beyond the cwd. Not sure whether this impacts the work you've done so far but hopefully a quick fix if so.

Btw, I'm going to be merging #1665 in the next couple of days which will allow for tool output to be folded. The first line of any tool output becomes the fold text so I'll start advising tool creators to think about how they output their text. Copilot-Chat is typically very concise in their summary, I assume owing to working in a small amount of screen real estate.

jinzhongjia commented on lua/codecompanion/strategies/chat/agents/tools/list_code_usages.lua:
```diff
@@ -0,0 +1,912 @@
+  local symbols = {}
+  for _, client in ipairs(clients) do
+    local params = { query = symbolName }
+    local result = client:request_sync(vim.lsp.protocol.Methods.workspace_symbol, params, CONSTANTS.LSP_TIMEOUT_MS)
```
All lsp requests should be asynchronous, and synchronization will have a clear sense of blockage in vue or rust projects.

lazymaniac replied:
Thanks for the comments. I use sync for simplicity, because I'm still testing some new ideas. Once I settle down with something I feel good about I will focus on making this code more production ready.
Async lsp is defiantly on my list and probably to split it into few files.

olimorris commented:
Amazing! Look forward to testing this out.

olimorris commented:
Firstly, thanks for such a killer PR. It's an unbelievable undertaking and the quality is supremely high. I had to pick my jaw up off the floor at just how quick it is from me making a dumb request to an LLM...

```
Can you use the @{list_code_usages} tool to investigate the code base for `add_buf_message` references?
```

...to seeing the LLM's analysis and response back in the chat buffer. The sheer amount of test coverage as well...wow!

olimorris commented on lua/codecompanion/strategies/chat/agents/tools/list_code_usages.lua:
```diff
@@ -0,0 +1,312 @@
+-- Main Tool Implementation
+-----------------------
+
+local function process_lsp_symbols_async(symbols, state, callback)
```
My only major critique of this PR is the lack of [annotations](https://luals.github.io/wiki/annotations/) - and even then this is a zero calorie suggestion. It just puts it at odds with all the other tools and the rest of the plugin.

lazymaniac replied:
Should be sorted out. (A)I added some extra comments for clarification.

olimorris replied:
I am totally stealing that line 😆. Can't believe I haven't seen that before.

olimorris commented on lua/codecompanion/strategies/chat/agents/tools/list_code_usages.lua:
```diff
@@ -0,0 +1,312 @@
+    end,
+  },
+  output = {
+    success = function(self, agent, cmd, stdout)
```
Previously, the output that was written to the chat buffer, took over everything:

<img width="1348" height="5422" alt="Image" src="https://github.com/user-attachments/assets/9aa92215-fd8c-46b5-9f13-3c8d4d6e1f2d" />

For my prompt, it's a large amount of text to scroll. By removing the initial line break, this can now be folded:

<img width="1348" height="1696" alt="Image" src="https://github.com/user-attachments/assets/21af39c1-7b6c-4834-98d1-49a8cb63db66" />

olimorris commented on lua/codecompanion/strategies/chat/agents/tools/list_code_usages.lua:
```diff
@@ -0,0 +1,312 @@
+      local symbol = self.args.symbol_name
+      local chat_message_content = string.format("Found usages of symbol: %s \n", symbol)
+
+      for operation, code_blocks in pairs(ListCodeUsagesTool.symbol_data) do
```
Something in here is also causing text to be formatted as a H2 header:

<img width="1348" height="1696" alt="Image" src="https://github.com/user-attachments/assets/7ce03c08-a630-431b-8380-a7f4737705de" />
